New discard per subject feature blog post (#792)
nats-io/nats-site: content/blog/new-per-subject-discard-policy.md

There are two parts to the exactly-once streaming QoS equation: 'publish exactly-once' which is done through server side message deduplication, and reliable consumption on the subscriber's side (a separate subject of its own) which is done through acknowledgements and re-deliveries. 

## Why would you need it?

Message deduplication is very useful to deal with two kinds of stream publishing failure scenarios:

1. Protecting against the failure scenarios where a client application 'fails' before it gets an acknowledgement of its publication being successful or not. In those scenarios, the application will typically re-publish the same message again upon re-connecting or restarting, thereby resulting in duplication of the message in the stream. This scenario can happen when the client application gets disconnected, crashes, when the server or client application process (or VM or container/pod) gets moved, updated, restarted, etc... This potential for message duplication also gets exacerbated when publication acks are handled asynchronously (such as when high throughput is desired).

> 2. Protecting against failure scenarios in stream processing pipelines (like Kafka Streams) where the event processing is happening in series of steps as messages representing the event being processed go through a number of streams (e.g. a message published on an upstream stream causes 1 or more messages to be published on one or more downstream streams). When there is a problem with one of the applications somewhere in the pipeline you can end up in a situation where some events can end up in a 'half-processed' state within the pipeline. In those scenarios, being able to recover the processing of those events simply by replaying original the event messages from the initial most stream *without having to worry about duplicate message publishing* (and its possible side effects in any of the steps of the pipeline as a result of the replay) becomes a very desirable functionality.

Author: You are right there's a problem with the sentence, corrected in the last commit

So, how do you actually use this new feature, and how does it work? Bear with me again as I first go through some prerequisite knowledge about stream limits and discard policies.

### Stream limits

JetStream allows you to set basic limits for a stream: the maximum number of messages in the stream, the maximum number of bytes, and you can specify the max amount of time you want the stream to keep messages (i.e. a TTL). On the other hand, Kafka and other distributed log systems allow you to adjust when log compaction kicks in to prune the tail end of the stream data.

### Discard policies

Because it is a data store, JetStream also allows you to specify a 'discard policy': what happens when a limit is breached because of a new publication. There are two discard policies:

- Discard `old` means that the stream will discard the oldest message in the stream in order to accept the new message being published while keeping the stream at the limit.

> This is probably the most common use-case: you want the stream to be a rolling window of data and yet control the size of the stream, new publications silently push out the oldest message in the stream. It is the only kind of discard policy that append only log streaming system with log compaction systems support, as they are meant to always accept new messages or block until they can.

Author: That JS has the concept of 'old' and 'new' discard policies, while log systems don't have the concept of a discard policy because they only can support the equivalent of the 'old' discard policy (because they always either accept the message to the stream of block until it can be accepted)

### Using Publish as an 'Update'

While it is outside the subject of this blog, I still want to mention that JetStream also has a 'compare and set' functionality (using the `nats.ExpectLastSequence()` publish option) which can be used to implement an equivalent to `update` in SQL.

## How does it work?
The ability to strictly and consistently enforce this limit (a key attribute for our use case) is possible because of a fundamental difference between JetStream and distributed log streaming systems: JetStream is 'subject-based addressing aware', it allows you to capture messages on multiple subjects into a single stream, rather than having a hard equivalency of exactly one topic per stream. Of course other streaming systems also allow the publishing application to provide a 'key' field along with the message being published to a topic, but the key is really meant to be only used for distribution (partitioning) purposes. But that is not the same thing as having the key(s) in the subject and then being able to use subject-based server-side (indexed) filtering consumers of the messages in a stream.

### Doesn't Kafka do this also?

> And indeed, if you know your Kafka, this may look to you just like the 'keep at least one message per key' log compaction feature, but there is a big difference: JetStream is 'exactly *n*' maximum number of messages per subject while log compaction only ensures that you will have 'at least 1' message per key.

Reviewer: ```suggestion
And indeed, if you have experience with Kafka, this may look to you just like the "keep at least one message per key" log compaction feature, but there is a big difference: JetStream is exactly *n* maximum number of messages per subject while log compaction only ensures that you will have *at least one* message per key.
```

> Besides the potential inefficiency (i.e. having to scan the whole to be sure to get the latest message for a specific key in the last message cache) you will not be able to use a streaming WAL functionality to enforce a strict per key limit, and to signal the client application publishing a message immediately when the publication would break the limit. This is why there is no concept of a `new` discard policy in distributed log streaming systems, they are designed and designed and meant for always accepting new messages or block until they can.

Reviewer: ```suggestion
Besides the potential inefficiency (i.e. having to scan the whole stream to be sure to get the latest message for a specific key in the last message cache) you will not be able to use a streaming WAL functionality to enforce a strict per key limit, and to signal the client application publishing a message immediately when the publication would break the limit. This is why there is no concept of a `new` discard policy in distributed log streaming systems, they are designed and optimized for always accepting new messages or block until they can.
```

## How do I play with it?
Congratulations, you have now reached the practical exercise walk through part of this blog and you can test this for yourself using the `nats` CLI tool!

> Prerequisite is access to a JetStream-enabled NATS service infrastructure (e.g. running `nats-server -js` locally) and having the `nats` CLI context pointing to it.

Reviewer: ```suggestion
A prerequisite is access to a JetStream-enabled NATS server (e.g. running `nats-server -js` locally) and having the `nats` CLI context pointing to it.
```

```
Let's do a JetStream publish of a message for id=1
```
nats req foo.1 one
``` 
Which is successful (here the message is assigned sequence number 1):
```
17:08:57 Sending request on "foo.1"
17:08:57 Received with rtt 284.958µs
{"stream":"dedup", "seq":1}
```
Let's publish another message, for id=2
```
nats req foo.2 one
```
Which is also successful (and assigned sequence number 2):
```
17:09:03 Sending request on "foo.2"
17:09:03 Received with rtt 267.584µs
{"stream":"dedup", "seq":2}
```
Now let's try to publish another message for id=1
```
nats req foo.1 two
```
Which fails, thereby ensuring message deduplication:
```
17:09:11 Sending request on "foo.1"
17:09:11 Received with rtt 192.709µs
{"error":{"code":503,"err_code":10077,"description":"maximum messages per subject exceeded"},"stream":"dedup","seq":0}
```
At this point we can for example remove the existing message for id=2 using its sequence number (2 in this example)
```
nats stream rmm dedup 2
```
And attempt to publish a new message for id=2
```
nats req foo.2 two
```
Which works:
```
17:09:52 Sending request on "foo.2"
17:09:52 Received with rtt 318.125µs
{"stream":"dedup", "seq":3}
```

## Closing thoughts

> To resume: the newly introduced discard policy `new` for per subject limit now enables the ability to have an 'infinite deduplication window' to avoid duplicated publications that can otherwise happen to some failure scenarios.

Reviewer: ```suggestion
To review, the new discard policy *discard-new-per-subject* enables the ability to have an 'infinite deduplication window' to avoid duplicated publications that can otherwise happen to some failure scenarios.
```

> This new ability is specific to JetStream due to its fundamental functionality differences compared to other distributed streaming systems: it is subject based addressing capable and has the features of a proper data store (rather than the features of an append-only log with compaction). The functionality is built-in to NATS server itself rather than something you have to implement and deploy using something additional (e.g. Kafka Streams processes) and requiring an external data store (e.g. Redis). 

Reviewer: ```suggestion
This new ability is specific to JetStream due to its fundamental functional differences compared to other distributed streaming systems: it is subject-based addressing capable and has the features of a proper data store (rather than the features of an append-only log with compaction). The functionality is built-in to NATS server itself rather than something you have to implement and deploy separately (e.g. Kafka Streams processes) and requires an external data store (e.g. Redis). 
```

Ensuring message deduplication is also not the only use for this new feature: it can effectively be used for any kind of (optimistic) distributed processing concurrency access control (e.g. forms of locks, semaphores, logic gating, etc...), and you can combine it with message TTL to automatically clear the lock after some time. Other use cases include:

> - n Workers need to perform an action at most once/twice/thrice per period

Reviewer: ```suggestion

- N workers need to perform an action at most once/twice/thrice per period
```

> - Out of 1000s of workers a subset only n should be able to do something concurrently

Reviewer: ```suggestion
- Given 1000s of workers on an N subset should be able to do something concurrently
```
